qir/qat/StaticResourceComponent/ReplaceQubitOnResetPass.cpp

```diff
@@ -0,0 +1,223 @@
+                if (call_instr)
+                {
+                    auto f = call_instr->getCalledFunction();
+                    if (f == nullptr)
+                    {
+                        continue;
+                    }
+
+                    auto call_name = static_cast<std::string>(f->getName());
+                    if (call_name == "__quantum__qis__reset__body")
+                    {
+                        auto op = call_instr->getOperand(0);
+                        auto it = access_map.find(op);
+
+                        if (it != access_map.end())
+                        {
+                            qubits_mapping[it->second.index] = next_qubit_index++;
+                            to_remove.push_back(&instr);
+                        }
+                    }
+                }
+
+                for (uint64_t i = 0; i < instr.getNumOperands(); ++i)
+                {
+                    auto op = instr.getOperand(i);
+
+                    // In case we already did the mapping, we skip to next instruction
+                    if (already_replaced.find(op) != already_replaced.end())
+                    {
+                        continue;
+                    }
+
+                    // We only perform the mapping for constant integer casts. Anything else
+                    // cannot be remapped.
+                    auto it = access_map.find(op);
+                    if (it != access_map.end())
+                    {
+                        auto     n              = it->second.index;
+                        auto&    type           = it->second.type;
+                        uint64_t remapped_index = 0;
+
+                        // Getting remapped index based on resource type
+                        if (type == Qubit)
+                        {
+                            if (qubits_mapping.find(n) != qubits_mapping.end())
+                            {
+                                remapped_index = qubits_mapping[n];
+                            }
+                            else
+                            {
+                                if (logger_)
+                                {
+                                    logger_->setLocationFromValue(op);
+                                    logger_->internalError("Original qubit index not found.");
+                                }
+                                else
+                                {
+                                    throw std::runtime_error("Original qubit index not found.");
+                                }
+                                continue;
+                            }
+                        }
+                        else if (type == Result)
+                        {
+                            if (results_mapping.find(n) != results_mapping.end())
+                            {
+                                remapped_index = results_mapping[n];
+                            }
```

Comment:
To make sure I understand, this handling of result remapping is speculative right now should results need some handling in the future, correct? It looks like only qubits get remapped and only when they go through reset. While the infrastructure here seems like reasonable future proofing, since the file and the object are both named `ReplaceQubitOnResetPass` does it make sense to handle anything other than qubits here?

Reply:
This is theoretical for now as we do not have an instruction that causes the allocation of a new result although that may or may not be needed. This was one of the things where I wanted to ask you about partner requirements? Other than that, it is merely there for symmetry at this stage.